src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -170,7 +171,13 @@ String getClassNameInternal() {
         cacheKey = new CityHash128Key(sql);
 
         // Parse or fetch SQL metadata from cache.
-        ParsedSQLCacheItem parsedSql = getOrCreateCachedParsedSQL(cacheKey, sql);
+        ParsedSQLCacheItem parsedSql = getCachedParsedSQL(cacheKey, sql);
+        if (null != parsedSql) {
+            isExecutedAtLeastOnce = true;
+        }
+        else {
+            parsedSql = parseAndCacheSQL(cacheKey, sql);
+        }
```

Comment:
Providing a real world example from just yesterday...

I was trying to run the ``oltpbench/epinions`` benchmark against my ``speedify`` branch, but I wasn't seeing any performance gain at all.  Zero.  So, I added some logging to see what was going on...

The code benchmark code does this:
 * Grab a connection from the pool (in a *try-with-resources* block)
 * Prepare a statement (``conn.prepareStatement(sql)``
 * Set parameters and execute
 * Return the connection to the pool (exit the ``try``)
 * Rinse and repeat

No *single* ``PreparedStatement`` is *ever* executed more than once.  So, no *sp_prepsql* was occurring, only *sp_executesql*.

Then I realized ... that is what my company's application, and most Java web applications, so in 90% of the cases.
 * The user types a search, enters an order, etc. and then clicks "go" (or "save", or "search")
 * The web request comes in, grab a connection from the pool
 * Prepare a statement
 * Set parameters and execute
 * Close the connection (back to the pool), and return results back to the user

Again, individual PreparedStatement instances are rarely executed multiple times, so no *sp_prepsql*.  However, there are thousands of users doing this!  That SQL needs to be prepared!

And a user of SQL Server JDBC shouldn't have to force prepare-on-first-execute across the board.  [But you can see](https://github.com/brettwooldridge/oltpbench/blob/hibernate4/config/sample_epinions_config.xml#L7), in order to actually realize *any* of the optimizations we've been making on this branch, that is exactly what I had to do.

And not only for the ``epinions`` benchmark, but also the ``jpab`` [benchmark](https://github.com/brettwooldridge/oltpbench/blob/hibernate4/src/META-INF/persistence.xml#L28).  Surveying the suite of the ~30 benchmarks in that suite, developed by different individuals, nearly all of them require that.